pkg/workflows/testutils/runner.go

```diff
@@ -34,15 +49,61 @@ func (r *Runner) Run(factory *workflows.WorkflowSpecFactory) error {
 		return err
 	}
 
-	// TODO https://smartcontract-it.atlassian.net/browse/KS-442, implement this function
-	_ = spec
-	return errors.New("TODO https://smartcontract-it.atlassian.net/browse/KS-442")
+	if err = r.ensureGraph(spec); err != nil {
+		return err
+	}
+
+	r.setupSteps(factory, spec)
+
+	return r.walk(spec, workflows.KeywordTrigger)
+}
+
+func (r *Runner) ensureGraph(spec workflows.WorkflowSpec) error {
+	g, err := workflows.BuildDependencyGraph(spec)
+	if err != nil {
+		return err
+	}
+
+	if len(g.Triggers) != 1 {
+		return fmt.Errorf("expected exactly 1 trigger, got %d", len(g.Triggers))
+	}
+
+	edges, err := g.Edges()
+	if err != nil {
+		return err
+	}
+
+	for _, edge := range edges {
+		r.dependencies[edge.Target] = append(r.dependencies[edge.Target], edge.Source)
+	}
+
+	r.am, err = g.AdjacencyMap()
+	return err
+}
+
+func (r *Runner) setupSteps(factory *workflows.WorkflowSpecFactory, spec workflows.WorkflowSpec) {
+	for _, step := range spec.Steps() {
+		if step.Ref == "" {
+			step.Ref = step.ID
+		}
+
+		r.idToStep[step.Ref] = step
+		if run := factory.GetFn(step.Ref); run != nil {
```

Comment:
This is really unobvious IMO -- can you add a comment here to explain what it's doing?

Eg. it seems that GetFn() returns nil if the step isn't custom, is that right?